File: api/ponto.ts
import * as playwright from 'playwright-aws-lambda';

// @ts-ignore
import * as chromiumBinary from '@sparticuz/chromium';
// @ts-ignore
// const chromiumBinary = require('@sparticuz/chromium');

import { chromium } from 'playwright-core';

function getPassword(username: string): string {
    const password = process.env[username] || process.env[username.replaceAll('.', '_')];
    return password || '';
}

export async function registrarPonto(username: string, url: string, latitude?: string, longitude?: string) {
    const agora = new Date().toISOString();
    try {
        // const browser =  await playwright.launchChromium();

        const executablePath = await chromiumBinary.executablePath();

        // launch browser with external Chromium
        const browser = await chromium.launch({
            args: chromiumBinary.args,
            executablePath: executablePath,
            headless: true,
        });

        const context = await browser.newContext({
            geolocation: {
                latitude: latitude,
                longitude: longitude,
            },
            permissions: ['geolocation'],
        });
        const page = (await context.newPage()) as any;

        await page.route('**/*', (route: any) => {
            if (
                ['image', 'stylesheet', 'font'].includes(
                    route.request().resourceType()
                )
            ) {
                route.abort();
            } else {
                route.continue();
            }
        });

        await page.evaluate((o?: { lat?: string | number, lon?: string | number }) => {
            // @ts-ignore
            (navigator as any).geolocation.getCurrentPosition = function (success: any) {
                success({
                    coords: {
                        latitude: o?.lat,
                        longitude: o?.lon,
                        accuracy: 100
                    }
                });
            };
        }, { latitude, longitude });

        await page.goto(url);

        await page.waitForSelector('#usuario');
        await page.waitForSelector('#senha');

        console.log('Preenchendo o formulário...');

        const password = getPassword(username);

        await page.fill('#usuario', username);
        await page.fill('#senha', password); // TODO: Obter dinamicamente

        console.log("username", username);
        console.log("password", password);


        await page.keyboard.press('Enter');

        await page.waitForSelector('#btnRegistrar');
        await page.click('#btnRegistrar');

        // print page html
        const html = await page.content();
        console.log(html);
        const text = 'com sucesso.';
        // check if this text is appears in the page after clicking the button
        const confirmation = await page.waitForSelector('div.alert.alert-success.growl-animated', { timeout: 10000 })
            .then((el: any) => el.innerText)
            .catch(() => {
                console.log('Erro ao registrar ponto 1');
                // wait for the text to appear anywhere in the page
                return page.waitForSelector(`text=${text}`, { timeout: 10000 })
                    .then((el: any) => el.innerText)
                    .catch(async () => {
                        // get error message
                        console.log('Erro ao registrar ponto 2');
                        const geoError = await page.$('div.alert.alert-danger.growl-animated');
                        if (geoError) {
                            const errorText = await geoError.innerText();
                            console.log('Erro de geolocalização:', errorText);
                            return errorText;
                        }
                        return 'Erro ao registrar ponto';
                    }
                    );
            }
            );
        console.log('confirmation', confirmation);

        if (confirmation.includes('Erro')) {
            console.log('Erro');
            return {
                usuario: username,
                hora: agora,
                mensagem: `Erro ao registrar ponto: ${confirmation}`,
                confirmacao: confirmation
            };
        }

        // console.log('div.alert.alert-success.growl-animated');

        // // wait for the element to be visible
        // // <div data-growl="container" class="alert alert-success gro wl-animated animated fadeInDown" role="alert" data-growl-position="top-right" style="position: fixed; margin: Opx; z-index: 1031; display: inline-block; top: 20px; right: 20 px;">@</div>
        // const confirmation = await page.waitForSelector('div.alert.alert-success.growl-animated', { timeout: 10000 })
        //     .then((el: any) => el.innerText)
        //     .catch(() => {
        //         return 'Erro ao registrar ponto';
        //     });
        // console.log('confirmation', confirmation);
        // await page.waitForTimeout(1000);

        await browser.close();

        return {
            usuario: username,
            hora: agora,
            mensagem: `Registrando ponto às ${agora}`,
            confirmacao: confirmation
        };

    } catch (error: any) {
        return {
            usuario: username,
            hora: agora,
            mensagem: `Erro ao registrar ponto: ${error.message}`,
            error: error.stack
        };
    }
}

File: api/ponto.py
import os
from dotenv import load_dotenv
import ssl
import time
import warnings
from datetime import datetime
from pytz import timezone
 
load_dotenv()
 
# Configurações de SSL
ssl._create_default_https_context = ssl._create_unverified_context
 
# URL
URL = os.getenv('URL')
# Chave da API
google_maps_api_key = os.getenv('GOOGLE_MAPS_API')

def get_password(username):
    """
    Função para obter a senha do usuário a partir de um arquivo .env.
    """
    load_dotenv()
    password = os.getenv(username)
    if password is None:
        # check if replacing '.' with '_'
        # if username has '.' replace with '_'
        username = username.replace('.', '_')
        password = os.getenv(username)

    if password is None:
        raise ValueError(f"Senha não encontrada para o usuário: {username}")
    return password
 
def registrar_ponto(username, url):
    password = get_password(username)
    agora = datetime.now().strftime('%Y-%m-%d %H:%M:%S')
    print(f"Registrando ponto às {agora} (Dia da semana: {datetime.now().strftime('%A')})")
    # Configuração do Chrome
    chrome_options = Options()
    chrome_options.add_argument("--no-sandbox")
    chrome_options.add_argument("--disable-dev-shm-usage")
 
    # Inicializar o WebDriver
    driver = webdriver.Chrome(service=Service(ChromeDriverManager().install()), options=chrome_options)
    try:
        # Injetar a API Key do Google Maps
        driver.execute_script(f"""
            var script = document.createElement('script');
            script.src = 'https://maps.googleapis.com/maps/api/js?key={google_maps_api_key}&libraries=places';
            document.head.appendChild(script);
        """)
 
        # Definir a geolocalização
        driver.execute_script("""
            navigator.geolocation.getCurrentPosition = function(success) {
                var position = {
                    coords: {
                        latitude: -15.821305,
                        longitude: -47.893889,
                        accuracy: 100
                    }
                };
                success(position);
            };
        """)
 
        # Abrir a URL
        if url is None:
            url = URL
        driver.get(url)
        time.sleep(2)
 
        # Esperar até que o campo de nome de usuário esteja presente
        username_input = WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.ID, 'usuario'))
        )
        password_input = WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.ID, 'senha'))
        )
        # Inserir credenciais
        username_input.clear()
        username_input.send_keys(username)
        password_input.clear()
        password_input.send_keys(password)
 
        # Pressionar ENTER para enviar o formulário
        password_input.send_keys(Keys.RETURN)
        time.sleep(2)  # Aguardar resposta do login
 
        # Esperar até que a página de destino carregue
        WebDriverWait(driver, 10).until(
            EC.presence_of_element_located((By.ID, 'btnRegistrar'))
        )
        print("Página de destino carregada.")
 
        # Esperar até que o botão de registrar ponto esteja visível e clicável
        punch_button = WebDriverWait(driver, 10).until(
            EC.element_to_be_clickable((By.ID, 'btnRegistrar'))
        )
        print("Botão de registrar ponto disponível.")
 
        # Tentar clicar no botão de registrar ponto
        punch_button.click()
        print("Botão de registrar ponto clicado.")
 
        # Verificar se uma mensagem de confirmação aparece usando a classe CSS
        confirmation_message = WebDriverWait(driver, 20).until(
            EC.visibility_of_element_located((By.CSS_SELECTOR, "div.alert.alert-success.growl-animated"))
        )
        print("Ponto registrado com sucesso!")
        # retornar objeto de confirmação com usuario, hora e mensage de confirmação
        driver.quit()
        return {
            "usuario": username,
            "hora": agora,
            "mensagem": f"Registrando ponto às {agora} (Dia da semana: {datetime.now().strftime('%A')})",
            "confirmacao": confirmation_message.text
        }
 
    except Exception as e:
        print(f"Ocorreu um erro ao registrar o ponto: {e}")
        # driver.save_screenshot(f'erro_{agora}.png')
        # print("Screenshot capturada.")
        # retornar objeto de erro com usuario, hora e mensagem de erro
        driver.quit()
        return {
            "usuario": username,
            "hora": agora,
            "mensagem": f"Erro ao registrar ponto: {e}",
            "error": e
        }
    finally:
        print("Finalizando o WebDriver...")
